Min # Vendingmachine step4

undervineg opened the pull request.

undervine committed:
Fix a bug that occurs when pop beverages

When pop a beverage, StockManager records just the last beverage of inventory.
Now it records the exact beverage selected by users.

undervine committed:
Modify fullSupply() func in VendingMachine to get supply count from users

undervine committed:
Define 'Activate' protocol #17

Add perform(), showResult() interfaces.

undervine committed:
Add AdminMode, UserMode struct that conform ActivateMode protocol #19

Inject dependency on VendingMachine's Managable, UserServable protocol types respectively when creating objects.

undervine committed:
Add Purchasable protocol

undervine committed:
Implement prompt() of UserMode, to use in main() #20

undervine committed:
Add user menu to choose AdminMode or UserMode #21

Add user menu to choose mode, and implement prompt() to return ActivateMode type instance according to the chosen mode.

undervine committed:
Add typealias 'MenuType' to VendingMachine's 'Menu'

Because Machine protocol has associatedtype 'MenuType', conforming class 'VendingMachine' should have it's typealias.
Add typealias 'MenuType' to VendingMachine's 'Menu' and use it in AdminMode, UserMode class.

undervine committed:
Move checkTheStock() from UserServable to Machine

undervine committed:
Add struct 'OutputView' and implement showResult() to print the final result #22

undervine committed:
Add Bool property 'isManagerRemoved' to separate just removed product by manager from purchased product by consumer

undervine committed:
Modify parsecond parameter label of updateStock() from 'isPurchased' to 'isRemoved'

undervine committed:
Modify functions

Modify prompt() to return optional type property.
Modify perform() to get paramter of user input arguments, and return result of optional string type to separate print() responsibility from it to OutputView.

undervine committed:
Add case 'exit' to prompt()

undervine committed:
Modify main()

Add while statment to call prompt() of ActivatedMode object repeatedly.
Add perform() of ActivatedMode object, and give the result to OutputView to print it.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -42,28 +59,74 @@ final class VendingMachine: Machine, Sequence {
+            var beverage = Beverage()
+            switch self {
+            case .strawberryMilk: beverage = StrawBerryMilk("서울우유", 200, 1000, "날마다딸기우유", Date(timeIntervalSinceNow: 0), Date(timeIntervalSinceNow: 60*60*24*7), 210, Menu.strawberryMilk, manufacturerCode: 1001, packingMaterial: "종이")
+            case .bananaMilk: beverage = BananaMilk("서울우유", 200, 1000, "날마다바나나우유", Date(timeIntervalSinceNow: 0), Date(timeIntervalSinceNow: 60*60*24*7), 220, Menu.bananaMilk, manufacturerCode: 1001, packingMaterial: "종이")
```
enum 타입에 관련된 beverage를 generate() 할 때, 인스턴스 생성은 여기가 아니라 Beverage 객체가 팩토리 패턴으로 구현했으면 좀 더 깔끔했을까요? 
- 예를 들어 StrawBerryMilk 초기화 메소드 인자값들이 많으니까, generate 전용으로 편리한 convenience 메소드를 추가해서 부르면 여기 있는 기본값을 채워주도록 하는 방법도 있습니다.

godrm replied:
Menu 열거형에서 generate()를 빼는 것보다 내부 구현 방식을 Beverage 객체가 Menu 타입을 받아서 만들어주는 방식도 깔끔할 것 같았습니다. 참고하세요.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -42,28 +59,74 @@ final class VendingMachine: Machine, Sequence {
+            case .fanta: beverage = FantaSoftDrink("코카콜라컴퍼니", 350, 2000, "환타", Date(timeIntervalSinceNow: 0), Date(timeIntervalSinceNow: 60*60*24*30*7), 300, Menu.fanta, carbonContent: 40)
+            case .top: beverage = TopCoffee("맥심", 200, 2200, "티오피", Date(timeIntervalSinceNow: 0), Date(timeIntervalSinceNow: 60*60*24*14), 240, Menu.top, caffeineLevels: 20, isHot: true, isSweetened: true)
+            case .cantata: beverage = CantataCoffee("롯데칠성음료", 200, 2200, "칸타타", Date(timeIntervalSinceNow: 0), Date(timeIntervalSinceNow: 60*60*24*14), 250, Menu.cantata, caffeineLevels: 10, isHot: false, isSweetened: true)
+            case .georgia: beverage = GeorgiaCoffee("코카콜라", 200, 2200, "조지아", Date(timeIntervalSinceNow: 0), Date(timeIntervalSinceNow: 60*60*24*2), 100, Menu.georgia, caffeineLevels: 25, isHot: true, isSweetened: false)
```
여기에 사용된 용량, 가격, 날짜, 칼로리 등 하드코딩하는 값 중에서 의미있는 코드로 대체 가능한건 선언해서 사용하는 게 좋습니다.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -42,28 +59,74 @@ final class VendingMachine: Machine, Sequence {
-            supply(beverageType: menu, 10)
+    // 모든 메뉴의 재고를 count개씩 자판기에 공급.
+    func fullSupply(_ count: Int) {
+        for menu in MenuType.allValues {
```
MenuType을 활용한 아이디어는 좋네요.

godrm commented on VendingMachine/VendingMachine/View/OutputView.swift:
```diff
@@ -2,38 +2,14 @@
-    init(_ vendingMachine: VendingMachine) {
-        self.vendingMachine = vendingMachine
+    static func showResult(_ result: String) {
+        print(result)
```
로직 코드에 출력 메시지를 만드는 코드가 다 들어가서 OutputView 기능이 많이 사라졌네요.
지금은 콘솔 출력이라서 이 구조도 괜찮은데, iOS 프로그래밍을 고려하면 좋은 방법은 아닙니다.
출력하는 View에 모든걸 만들어서 넘기기 보다는, 필요한 데이터만 넘기고 View에서 만드는 방식이 더 나을 수 있습니다. 참고하세요.

godrm commented on VendingMachine/VendingMachine/StockManager.swift:
```diff
@@ -21,27 +21,8 @@ class StockManager {
-        guard let changedBeverage = change else { return }
-        stock = stock.update(forKey: changedBeverage.menuType, isAdded)
+    func updateStock(_ recentChanged: Beverage, isRemoved: Bool) {
+        stock = stock.update(forKey: recentChanged.menuType, !isRemoved)
```
주로 비교하는 값이 음료를 추가한 경우라면 isRemoved로 선언하고 !(부정) 시킬 필요없이
isNewArrival 처럼 긍정형으로 바꾸고 ! 를 생략하는게 읽기 편합니다.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -9,13 +9,15 @@
+        self.isManagerRemoved = false
         // 장부기록, 돈관리의 책임을 위임.
         self.stockManager = StockManager(self)
         self.moneyManager = MoneyManager(self)
```
VendingMachine 에 포함된 StockManager와 MoneyManager 가 VendingMachine에 의존성을 갖는데,
이것도 inventory에 있는 Beverage에 접근하는 프로토콜에 의존하도록 만들수 있었을 것 같네요.

godrm reviewed (commented):
전체적인 구조에서 프로토콜과 제네릭 연관타입을 활용한 부분이 돋보입니다. 
프로토콜 타입과 타입이 중첩되면서 구조적으로 학습하는데 조금 복잡하지만, 그만큼 의미있는 방식입니다. 
아래 항목들을 확인해보시고, 내일 마무리하면 될 것 같습니다.

undervine committed:
Modify label 'isPurchased:' to 'isNewArrival:' for easy read

undervine committed:
Make convenience init() on each beverage classes

undervine committed:
Add data dictionary for beverage instances

undervine committed:
Add 'Product' protocol to use classes like Beverage generally

undervine committed:
Replace all 'Beverage' type to 'Product' type

undervine committed:
Remove all of concrete types from protocols, such as 'Beverage'

undervine committed:
Fix bugs that occurs according to removing concrete types from protocols

undervine committed:
Separate mutatable data from initializer, instead, make dictionary to store them

Add dictionaries 'stringData', 'intData' in BeverageData.swift.
Access data from convenience initializer of each beverage type classes.

undervine committed:
Try to refactor codes not to return final output String, but to return just required data

Try to add struct that stores result data.
Try to call the struct initializer in perform- functions of AdminMode and UserMode class to return only data required to print.
But, 'Command failed due to signal: Illegal instruction: 4' error occured.
It's because of the new made struct (Result struct).
Have to remove the new made Result struct.

undervineg commented:
- isRemoved  -> isNewArrival 변경
- StockManager와 MoneyManager 을 VendingMachine과 Beverage를 사용하는 형태에서 MachineType과 ProductType을 사용하는 형태로 변경 (제네릭 활용) : 프로토콜로 한다는 걸.. 헷갈렸는지 제네릭으로 해버렸네요 ㅎㅎ
- Beverage를 상속받는 클래스들에 convenience 생성자 추가
- VendingMachine.Menu 열거형 내부의 generate()를 바깥으로 빼려고 했으나, 열거형 내부에서 generate()가 필요한 부분이 있어 놔뒀습니다. 옮기면 너무 복잡해지더라구요..
- 용량, 가격, 날짜, 칼로리 등 하드코딩한 부분을 딕셔너리 데이터로 분리하여 사용
- OutputView에 String을 넘겨서 프린트만 하는 부분: Result 구조체를 만들어서 필요한 데이터만 넘긴 후 OutputView에서 출력해주려고 시도했으나, 'Command failed due to signal: Illegal instruction: 4' 에러가 떠서 제거함.

godrm commented on VendingMachine/VendingMachine/Beverage/Coffee/CantataCoffee.swift:
```diff
@@ -14,6 +14,11 @@ class CantataCoffee: Coffee {
     }
+
+    convenience init(_ menu: VendingMachine.Menu) {
+        self.init(stringData["brands"]![menu]!, intData["volume"]![menu]!, intData["price"]![menu]!, stringData["productName"]![menu]!, Date(timeIntervalSinceNow: 0), Date(timeIntervalSinceNow: TimeInterval(intData["expirationDate"]![menu]!)), intData["calories"]![menu]!, menu, caffeineLevels: intData["caffeineLevels"]![menu]!, isHot: false, isSweetened: true)
```
Dictionary 형태로 바꾸면서 다양한 값을 미리 넣어놓을 수 있었는데, 키값들이 다시 하드코딩이 됐네요 :) 
키값 문자열도 하드코딩을 피할 수 있는 방법은 뭐가 있을까요? 고민해보세요 (수정할 필요는 없습니다)

godrm commented on VendingMachine/VendingMachine/MoneyManager.swift:
```diff
@@ -8,13 +8,13 @@
 
-class MoneyManager {
-    private weak var vendingMachine: VendingMachine!
+class MoneyManager<MachineType: Machine> {
```
제네릭 타입으로 추상화를 하는 경우는 동일한 알고리즘을 여러 타입에서 구현하는 경우가 적합합니다. 
프로토콜을 써야 하는 경우와 제네릭을 써야 하는 경우를 기준을 구분해보세요.

godrm commented:
이번 미션에서 연습한 내용들을 꼭 정리해보세요.
프로토콜과 연관타입, 그리고 제네릭 패턴은 코드를 단순화 시켜주지만, 
잘못하면 코드 복잡도를 높이기 때문에 주의가 필요합니다.
레벨2 졸업을 축하합니다. 레벨3에 도전하세요!
